src/meta/src/hummock/manager/compaction/mod.rs

```diff
@@ -728,6 +729,46 @@ impl HummockManager {
                 }
             }
 
+            // Filter out the table that has a primary key prefix watermark.
+            let table_id_with_pk_prefix_watermark: HashSet<_> = self
+                .metadata_manager
+                .catalog_controller
+                .get_table_by_ids(
+                    version
+                        .latest_version()
+                        .table_watermarks
+                        .keys()
+                        .map(|id| id.table_id() as _)
+                        .collect(),
+                )
+                .await
+                .map_err(|e| Error::Internal(e.into()))?
+                .into_iter()
+                .filter_map(|table| {
+                    // pk prefix watermark.
+                    if table.clean_watermark_index_in_pk.is_none()
+                        || table.clean_watermark_index_in_pk.unwrap() == 0
+                    {
+                        Some(TableId::from(table.get_id()))
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            let table_watermarks = version
+                .latest_version()
+                .table_watermarks
+                .iter()
+                .filter_map(|(table_id, table_watermarks)| {
+                    if table_id_with_pk_prefix_watermark.contains(table_id) {
```

Comment:
We already have a `WaterMarkType` [define](https://github.com/risingwavelabs/risingwave/blob/3500061180b4eafb23dbeed1c41c0ee7d00ae97d/src/storage/hummock_sdk/src/table_watermark.rs#L412) in the version, why don't we just use that to filter out table with non pk prefix watermark?

Reply:
Also, if we filter out non pk prefix watermark here, how can compactor retrieve the non pk prefix watermark? Based on the logic [here](https://github.com/risingwavelabs/risingwave/blob/3500061180b4eafb23dbeed1c41c0ee7d00ae97d/src/storage/src/hummock/compactor/compactor_runner.rs#L249), it seems that we rely on the fact that non pk prefix watermark is present in the compact task.

Reply:
Good catch , we should filter the watermark by WaterMarkType directly.

And, the filtered results are only passed to the picker, while all relevant watermarks are passed to the compactor (pk or non-pk).